src/core/public/context/context_service.mock.ts

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ContextSetup, ContextService } from './context_service';
+
+const createSetupContractMock = () => {
+  const setupContract: jest.Mocked<ContextSetup> = {
+    createContextContainer: jest.fn(),
+  };
+  return setupContract;
+};
+
+type ContextServiceContract = PublicMethodsOf<ContextService>;
+const createMock = () => {
```

Comment:
I know I keep beating a dead horse on this one, but do we know **why** TypeScript doesn't infer this correctly?

```ts
const createMock = (): jest.Mocked<ContextServiceContract> => ({
  setup: jest.fn().mockReturnValue(createSetupContractMock()),
});
```

It seems strange that we need to create temporary throw-away variables to satisfy some problem with typing here.

Reply:
I think @restrry knows more. I believe it's a fixable problem in DefinitelyTyped but not sure if it's been fixed yet?

Reply:
jest doesn't know which function it mocks. if we want to have type safety, we should provide types manually
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L120-L127
There are several options
- define a type manually for every function
```typescript
type Arguments<T> = T extends (...args: infer U) => any ? U : never;
//...
const mock = {
    start: jest
      .fn<
        ReturnType<typeof createStartContractMock>,
        Arguments<typeof createStartContractMock>
      >()
      .mockReturnValue(createStartContractMock),
}
```
- or define a result type
```typescript
const mock: jest.Mocked<ContextServiceContract> = {
    start: jest.fn(),
}
// and now, knowing the result type, jest can check whether a passed value satisfies the type
mock.start.mockImplementation(createStartContractMock);
```


when you declare a type as:
```typescript
const createMock = (): jest.Mocked<ContextServiceContract> => ({
  setup: jest.fn().mockReturnValue(createSetupContractMock()),
});
```
TS uses this declaration https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L123
and infers a type as `jest.fn<any, any>`
![2019-07-15_13-23-10](https://user-images.githubusercontent.com/3198181/61212926-da41da00-a703-11e9-8ef3-f6085cbc7583.png)
Thus, any value passed as `mockReturnValue` or `mockImplementation` satisfies this type.

probably there is a better solution I'm not aware of.